### PR TITLE
[CONTRACTS] Allow loop contracts annotated to goto statement

### PR DESCRIPTION
**Pull request by qinheping**
This PR enable loop contracts annotated to goto statement, which is we we do in Kani generated GOTO program. 

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/ansi-c/goto-conversion/goto_convert.cpp [hunk 1]
```diff
@@ -1020,11 +1020,20 @@ void goto_convertt::convert_loop_contracts(
...
+      if(check_side_effect)
+      {
+        throw incorrect_goto_program_exceptiont(
+          "Loop invariant is not side-effect free.",
+          code.find_source_location());
+      }
+      else
+      {
+        warning().source_location = code.find_source_location();
+        warning() << "Loop invariants is not side-effect-free." << eom;
```
Comment (tautschnig) [inline src/ansi-c/goto-conversion/goto_convert.cpp]: Nobody reads warnings. Can you describe the scenario where side effects would be ok?

Comment (qinheping) [reply src/ansi-c/goto-conversion/goto_convert.cpp]: According to the offline discussion. I will move all the checks and throw to contracts component. They should not be in the converter.

File: src/ansi-c/goto-conversion/goto_convert.cpp [hunk 2]
```diff
@@ -1020,11 +1020,20 @@ void goto_convertt::convert_loop_contracts(
   {
     if(has_subexpr(invariant, ID_side_effect))
     {
-      throw incorrect_goto_program_exceptiont(
-        "Loop invariant is not side-effect free.", code.find_source_location());
+      if(check_side_effect)
```
Comment (feliperodri) [inline src/ansi-c/goto-conversion/goto_convert.cpp]: Could we add test cases to cover these branches?

File: src/ansi-c/goto-conversion/goto_convert.cpp (hunk 1 above)
Comment (qinheping) [reply src/ansi-c/goto-conversion/goto_convert.cpp]: This PR now only allow loop contracts annotated to GOTO. Check for side effect will be in another PR.

File: src/ansi-c/goto-conversion/goto_convert.cpp (hunk 2 above)
Comment (qinheping) [reply src/ansi-c/goto-conversion/goto_convert.cpp]: This PR now only allow loop contracts annotated to GOTO. Check for side effect will be in another PR.

**Review by tautschnig — APPROVED** [review]

**Review by feliperodri — APPROVED** [review]

**Pull request merged**